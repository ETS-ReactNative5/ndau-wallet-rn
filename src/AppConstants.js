const HARDENED_CHILD_BIP_44 = 44
const ACCOUNT_CREATION_KEY_CHILD = 100
const VALIDATION_KEY = 2000
const NDAU_CONSTANT = 20036 // which is 0x4e44 in hex, which are the two letters ND in ASCII
const DERIVED_ROOT_YES = 'yes'
const DERIVED_ROOT_NO = 'no'
const DERIVED_ROOT_UNKNOWN = 'unknown'
const TESTNET_ADDRESS = 'tn'
const MAINNET_ADDRESS = 'nd'
const APP_LANGUAGE = 'en'

const NORMAL_MODE = 'normal'
const PASSWORD_RESET_MODE = 'password-reset'
const GENESIS_MODE = 'genesis'
const NEW_PASSWORD_MODE = 'new-password'

const TEMP_ID = 'temp-id'

const NDAU_EPOCH = '2000-01-01T00:00:00Z'
const QUANTA_PER_UNIT = 100000000
const RATE_DENOMINATOR = QUANTA_PER_UNIT * 10000

const KYC_ACCEPTED = 'ACCEPT'
const KYC_MANUAL_REVIEW = 'MANUAL_REVIEW'
const KYC_DENY = 'DENY'

const TOS_SETUP = 'termsOfServiceSetup'
const TOS_BUY = 'termsOfServiceBuyMode'

export default {
  HARDENED_CHILD_BIP_44,
  ACCOUNT_CREATION_KEY_CHILD,
  VALIDATION_KEY,
  NDAU_CONSTANT,
  DERIVED_ROOT_NO,
  DERIVED_ROOT_UNKNOWN,
  DERIVED_ROOT_YES,
  TESTNET_ADDRESS,
  MAINNET_ADDRESS,
  APP_LANGUAGE,
  NORMAL_MODE,
  PASSWORD_RESET_MODE,
  GENESIS_MODE,
  NEW_PASSWORD_MODE,
  TEMP_ID,
  NDAU_EPOCH,
  QUANTA_PER_UNIT,
  RATE_DENOMINATOR,
  KYC_ACCEPTED,
  KYC_MANUAL_REVIEW,
  KYC_DENY,
  TOS_BUY,
  TOS_SETUP
}
