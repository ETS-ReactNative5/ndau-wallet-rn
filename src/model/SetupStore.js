import { Alert } from 'react-native';
import AppConstants from '../AppConstants';

class SetupStore {
  constructor() {
    if (!SetupStore.instance) {
      SetupStore.instance = this;
    }

    this._userId = '';
    this._numberOfAccounts = 0;
    this._qrCode = '';
    this._encryptionPassword = '';
    this._entropy = '';
    this._recoveryPhrase = '';
    this._shuffledMap = [];
    this._shuffledWord = [];
    this._addressType = AppConstants.MAINNET_ADDRESS;
    this._walletName = '';
  }

  set userId(userId) {
    this._userId = userId;
    this._printData();
  }

  get userId() {
    console.debug(`SetupStore.userId ${this._userId}`);
    return this._userId;
  }

  set numberOfAccounts(numberOfAccounts) {
    this._numberOfAccounts = numberOfAccounts;
    this._printData();
  }

  get numberOfAccounts() {
    console.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`);
    return this._numberOfAccounts;
  }

  set qrCode(qrCode) {
    this._qrCode = qrCode;
    this._printData();
  }

  get qrCode() {
    console.debug(`SetupStore.qrCode ${this._qrCode}`);
    return this._qrCode;
  }

  set encryptionPassword(encryptionPassword) {
    this._encryptionPassword = encryptionPassword;
    this._printData();
  }

  get encryptionPassword() {
    console.debug(`SetupStore.encryptionPassword ${this._encryptionPassword}`);
    return this._encryptionPassword;
  }

  set entropy(entropy) {
    this._entropy = entropy;
    this._printData();
  }

  get entropy() {
    console.debug(`SetupStore.entropy ${this._entropy}`);
    return this._entropy;
  }

  set recoveryPhrase(recoveryPhrase) {
    this._recoveryPhrase = recoveryPhrase.slice();
    this._printData();
  }

  get recoveryPhrase() {
    console.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`);
    return this._recoveryPhrase;
  }

  set shuffledWords(shuffledWords) {
    this._shuffledWords = shuffledWords.slice();
    this._printData();
  }

  get shuffledWords() {
    console.debug(`SetupStore.shuffledWords ${this._shuffledWords}`);
    return this._shuffledWords;
  }

  set shuffledMap(shuffledMap) {
    this._shuffledMap = shuffledMap.slice();
    this._printData();
  }

  get shuffledMap() {
    console.debug(`SetupStore.shuffledMap ${this._shuffledMap}`);
    return this._shuffledMap;
  }

  toggleAddressType() {
    const oldAddressType = this._addressType;
    const newAddressType =
      this._addressType === AppConstants.MAINNET_ADDRESS
        ? AppConstants.TESTNET_ADDRESS
        : AppConstants.MAINNET_ADDRESS;

    Alert.alert(
      'Information',
      `Old address type was ${oldAddressType} which has been moved to ${newAddressType}`,
      [{ text: 'OK', onPress: () => { } }],
      { cancelable: false }
    );

    this._addressType = newAddressType;
    this._printData();
  }

  get addressType() {
    console.debug(`SetupStore.addressType ${this._addressType}`);
    return this._addressType;
  }

  set walletName(walletName) {
    this._walletName = walletName;
    this._printData();
  }

  get walletName() {
    console.debug(`SetupStore.walletName ${this._walletName}`);
    return this._walletName;
  }

  _printData = () => {
    // console.debug(`SetupStore.userId ${this._userId}`);
    // console.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`);
    // console.debug(`SetupStore.qrCode ${this._qrCode}`);
    // console.debug(`SetupStore.entropy ${this._entropy}`);
    // console.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`);
    // console.debug(`SetupStore.shuffledWords ${this._shuffledWords}`);
    // console.debug(`SetupStore.shuffledMap ${this._shuffledMap}`);
    // console.debug(`SetupStore.walletName ${this._walletName}`);
    // console.debug(`SetupStore.addressType ${this._addressType}`);
  };
}

const instance = new SetupStore();

export default instance;
