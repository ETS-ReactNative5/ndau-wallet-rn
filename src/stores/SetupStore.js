import { Alert } from 'react-native'
import AppConstants from '../AppConstants'
import LoggingService from '../services/LoggingService'

class SetupStore {
  constructor () {
    if (!SetupStore.instance) {
      SetupStore.instance = this
    }

    this._userId = ''
    this._numberOfAccounts = 0
    this._qrCode = ''
    this._encryptionPassword = ''
    this._entropy = ''
    this._recoveryPhrase = ''
    this._shuffledMap = []
    this._shuffledWord = []
    this._addressType = AppConstants.MAINNET_ADDRESS
    this._walletId = ''
  }

  set userId (userId) {
    this._userId = userId
  }

  get userId () {
    LoggingService.debug(`SetupStore.userId ${this._userId}`)
    return this._userId
  }

  set numberOfAccounts (numberOfAccounts) {
    this._numberOfAccounts = numberOfAccounts
  }

  get numberOfAccounts () {
    LoggingService.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`)
    return this._numberOfAccounts
  }

  set qrCode (qrCode) {
    this._qrCode = qrCode
  }

  get qrCode () {
    LoggingService.debug(`SetupStore.qrCode ${this._qrCode}`)
    return this._qrCode
  }

  set encryptionPassword (encryptionPassword) {
    this._encryptionPassword = encryptionPassword
  }

  get encryptionPassword () {
    LoggingService.debug(`SetupStore.encryptionPassword ${this._encryptionPassword}`)
    return this._encryptionPassword
  }

  set entropy (entropy) {
    this._entropy = entropy
  }

  get entropy () {
    LoggingService.debug(`SetupStore.entropy ${this._entropy}`)
    return this._entropy
  }

  set recoveryPhrase (recoveryPhrase) {
    this._recoveryPhrase = recoveryPhrase.slice()
  }

  get recoveryPhrase () {
    LoggingService.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`)
    return this._recoveryPhrase
  }

  set shuffledWords (shuffledWords) {
    this._shuffledWords = shuffledWords.slice()
  }

  get shuffledWords () {
    LoggingService.debug(`SetupStore.shuffledWords ${this._shuffledWords}`)
    return this._shuffledWords
  }

  set shuffledMap (shuffledMap) {
    this._shuffledMap = shuffledMap.slice()
  }

  get shuffledMap () {
    LoggingService.debug(`SetupStore.shuffledMap ${this._shuffledMap}`)
    return this._shuffledMap
  }

  toggleAddressType () {
    const oldAddressType = this._addressType
    const newAddressType =
      this._addressType === AppConstants.MAINNET_ADDRESS
        ? AppConstants.TESTNET_ADDRESS
        : AppConstants.MAINNET_ADDRESS

    Alert.alert(
      'Information',
      `Old address type was ${oldAddressType} which has been moved to ${newAddressType}`,
      [{ text: 'OK', onPress: () => {} }],
      { cancelable: false }
    )

    this._addressType = newAddressType
  }

  get addressType () {
    LoggingService.debug(`SetupStore.addressType ${this._addressType}`)
    return this._addressType
  }

  set walletId (walletId) {
    this._walletId = walletId
  }

  get walletId () {
    LoggingService.debug(`SetupStore.walletId ${this._walletId}`)
    return this._walletId
  }

  printData = () => {
    LoggingService.debug(`SetupStore.userId ${this._userId}`)
    LoggingService.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`)
    LoggingService.debug(`SetupStore.qrCode ${this._qrCode}`)
    LoggingService.debug(`SetupStore.entropy ${this._entropy}`)
    LoggingService.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`)
    LoggingService.debug(`SetupStore.shuffledWords ${this._shuffledWords}`)
    LoggingService.debug(`SetupStore.shuffledMap ${this._shuffledMap}`)
    LoggingService.debug(`SetupStore.walletId ${this._walletId}`)
    LoggingService.debug(`SetupStore.addressType ${this._addressType}`)
  }

  reset = () => {
    this._userId = ''
    this._numberOfAccounts = 0
    this._qrCode = ''
    this._encryptionPassword = ''
    this._entropy = ''
    this._recoveryPhrase = ''
    this._shuffledMap = []
    this._shuffledWord = []
    this._addressType = AppConstants.MAINNET_ADDRESS
    this._walletId = ''
  }
}

const instance = new SetupStore()

export default instance
