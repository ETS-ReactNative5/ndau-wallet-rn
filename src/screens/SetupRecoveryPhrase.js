import React, { Component } from 'react'
import { PixelRatio, NativeModules, Alert } from 'react-native'
import groupIntoRows from '../helpers/groupIntoRows'
import SetupStore from '../stores/SetupStore'
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen'
import AppConstants from '../AppConstants'
import LogStore from '../stores/LogStore'
import {
  SetupContainer,
  RecoveryPhraseConfirmationTextOnly
} from '../components/setup'
import FlashNotification from '../components/common/FlashNotification'
import { LargeButtons, ParagraphText } from '../components/common'

var _ = require('lodash')

const DEFAULT_ROW_LENGTH = 4

class SetupRecoveryPhrase extends Component {
  constructor (props) {
    super(props)
    this.state = {
      recoveryPhrase: []
    }

    this.boxWidth = '30%'
    this.boxHeight = '14%'
    this.rowLength = DEFAULT_ROW_LENGTH
    // if someone has cranked up the font use 1 row instead
    LogStore.log(`PixelRatio.getFontScale is ${PixelRatio.getFontScale()}`)
    if (PixelRatio.getFontScale() > 2) {
      this.rowLength = 1
      this.boxWidth = '100%'
      this.boxHeight = '30%'
      LogStore.log(
        `boxWidth: ${this.boxWidth} and boxHeight: ${this.boxHeight}`
      )
    }
    props.navigation.addListener('didBlur', FlashNotification.hideMessage)
  }

  componentDidMount = () => {
    this.generateRecoveryPhrase()
  }

  showExitApp () {
    Alert.alert(
      '',
      `Problem occurred validating the phrase, please contact Oneiro.`,
      [
        {
          text: 'Exit app',
          onPress: () => {
            RNExitApp.exitApp()
          }
        }
      ],
      { cancelable: false }
    )
  }

  generateRecoveryPhrase = async () => {
    const KeyaddrManager = NativeModules.KeyaddrManager
    const seeds = await KeyaddrManager.keyaddrWordsFromBytes(
      AppConstants.APP_LANGUAGE,
      SetupStore.entropy
    )
    const seedBytes = await KeyaddrManager.keyaddrWordsToBytes(
      AppConstants.APP_LANGUAGE,
      seeds
    )
    if (!_(seedBytes).isEqual(SetupStore.entropy)) {
      this.showExitApp()
    } else {
      LogStore.log(`the seedBytes and entropy are equal.`)
    }
    const recoveryPhrase = seeds.split(/\s+/g)
    this.setState({ recoveryPhrase: recoveryPhrase })

    // Shuffle the deck with a Fisher-Yates shuffle algorithm;
    // walks through the array backwards once, exchanging each value
    // with a random element from the remainder of the array
    let map = recoveryPhrase.map((e, i) => i)
    let arr = recoveryPhrase.slice()
    for (let i = arr.length - 1; i >= 0; i--) {
      let r = Math.floor(Math.random() * i)
      ;[map[i], map[r]] = [map[r], map[i]]
      ;[arr[i], arr[r]] = [arr[r], arr[i]]
    }
    this.shuffledWords = arr
    // turn array inside out
    this.shuffleMap = map.reduce((a, c, i) => {
      a[c] = i
      return a
    }, [])
  }

  showNextSetup = () => {
    SetupStore.recoveryPhrase = this.state.recoveryPhrase
    SetupStore.shuffledMap = this.shuffleMap
    SetupStore.shuffledWords = this.shuffledWords

    const { navigation } = this.props
    navigation.navigate('SetupConfirmRecoveryPhrase')
  }

  render () {
    // chop the words into DEFAULT_ROW_LENGTH-tuples
    const words = groupIntoRows(this.state.recoveryPhrase, this.rowLength)
    const styles = {
      rowTextView: {
        height: hp(this.boxHeight),
        width: wp(this.boxWidth)
      }
    }

    return (
      <SetupContainer {...this.props} pageNumber={15}>
        <ParagraphText>
          Please write down this phrase, in this exact order, and plan to store
          it in a secure location.
        </ParagraphText>
        <RecoveryPhraseConfirmationTextOnly
          words={words}
          rowTextView={styles.rowTextView}
        />
        <LargeButtons
          sideMargins
          bottom
          onPress={() => this.showNextSetup()}
          text='We will confirm on the next screen.'
        >
          I wrote it down
        </LargeButtons>
      </SetupContainer>
    )
  }
}

export default SetupRecoveryPhrase
