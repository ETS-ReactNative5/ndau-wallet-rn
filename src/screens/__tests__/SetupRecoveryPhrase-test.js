import React from 'react'
import { StyleSheet, NativeModules } from 'react-native'
import SetupRecoveryPhrase from '../SetupRecoveryPhrase'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import SetupStore from '../../model/SetupStore'
import AppConstants from '../../AppConstants'

const mockKeyaddr = () => {
  NativeModules.KeyaddrManager = {
    keyaddrWordsFromBytes: jest.fn((lng, ent) => {
      const lorem = 'lorem ipsum dolor sit amet consectetur adipiscing elit sed do eiusmod tempor incididunt ut labore et dolore magna aliqua ut enim ad minim veniam quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur excepteur sint occaecat cupidatat non proident sunt in culpa qui officia deserunt mollit anim id est laborum'.split(
        ' '
      )
      const Base64 = require('base-64')
      const data = Base64.decode(ent)
      // not a Serious Hashâ„¢
      let hash = data.split('').reduce((a, c, i) => {
        a += c.charCodeAt(0)
        return a
      }, 1)
      const start = hash % lorem.length
      return lorem.slice(start, start + Math.floor(data.length * 0.8))
    })
  }
}

const styles = StyleSheet.create({
  wizardText: {
    color: '#ffffff',
    fontSize: 20
  }
})

const navigator = {
  setStyle: () => {},
  toggleNavBar: () => {}
}

const makeComponent = () => (
  <SetupRecoveryPhrase navigator={navigator} parentStyles={styles} />
)

describe('testing SetupRecoveryPhrase...', () => {
  it('calls keyaddrWordsFromBytes as expected', () => {
    mockKeyaddr()
    SetupStore.entropy = 'dGVzdGluZ3dlc3Rpbmdh'
    this.wrapper = mount(makeComponent())
    this.wrapper.update()
    expect(
      NativeModules.KeyaddrManager.keyaddrWordsFromBytes.mock.calls
    ).toEqual([[AppConstants.APP_LANGUAGE, 'dGVzdGluZ3dlc3Rpbmdh']])
  })
})
