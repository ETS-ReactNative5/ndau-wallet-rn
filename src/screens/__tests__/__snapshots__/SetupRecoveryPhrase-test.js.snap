// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SetupRecoveryPhrase... renders correctly 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 26.5,
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <RCTScrollView
      keyboardShouldPersistTaps="always"
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 20,
              }
            }
          >
            Write this 12-word phrase down and store it in a secure location.
          </Text>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignContent": "center",
            "backgroundColor": "#4e957a",
            "borderRadius": 3,
            "height": 93.5,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "center",
              "color": "#ffffff",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          I wrote it down
        </Text>
      </View>
    </View>
  </View>
</View>
`;
